### ISSUE TITLE
Class Hierarchy/Responsibilities 

### ISSUE THREAD

**Issue by ryancrideout** [ryancrideout/Ryan_Professional_Development#32]
Okay what should be responsible for what? Here's how Ryan visualizes things:

```
Game
 | - Map
 |    | - MapTile
 |
 | - Character
```

**Game Class**
I think I'm okay with the game class being responsible for EVERYTHING.

The reason being that the game I feel is responsible for loading everything up, from the characters, maps, and everything else. I feel like the game should also do all of the calculations as well.

Part of my reasoning for having the game class be responsible for everything is because it just feels so intuitive. If I consider my interpretation of video games, I feel like game engines (which is what I imagine the game class to be in this case) are responsible for running the game as well as everything that goes with it. That, and if the game class weren't responsible for everything else then I don't know what would be responsible for the remaining elements.

**Map Class**
I am fine with Maps being responsible for MapTiles. That seems intuitive to me as a Map is composed up of MapTiles.

**MapTile Class**
MapTiles should keep track of who/what is on them, and nothing else. I don't think they should care about other MapTiles, or by extension, the Map.

One question that arises from this is, is the game class responsible for MapTiles? While Maps are responsible for MapTiles, I feel like the game will still have it's hand in the calculations that concern themselves with MapTiles. So in an indirect sense, the game class will be responsible for Map Tiles, but I think it'll still be the Map that keeps track of how many/what it's MapTiles are.

**Character Class**
Characters should NOT concern themselves with Map or MapTile information. A character will have a position, sure, but that's not going to be on the character to verify if that's a legitimate map coordinate. A MapTile will know if there's a character/entity occupying that space but that's the extent of it. It'll be up to the game to determine, "Hey that character is on a legitimate tile, and is not out of bounds of the map".

**Thoughts**
With this in mind, here are some ideas to keep this philosophy consistent in the code:
- The "render" method that is attached to the maps should instead be attached to the game.

- The more I think about, the more I can see an argument for the Game class being responsible for knowing if a MapTile is occupied or not. While it is convenient to have the MapTile keep track of it's occupants, if the Game class is truly handling all of the calculations then it's not a stretch to have it keep track of occupancy calculations. So I guess this means moving the "set_occupant" method from the MapTile class to the Game class?

- There's an idea to have the Game class to keep track of the character positions as well. While right now characters have coordinate information on them, I see the logic in having the Game keeping track of where the characters. After all, Characters don't really do anything with their coordinates, so maybe it doesn't make much sense to put those coordinates on the Characters.

**Comment by sepiariver** [comment]
Sometimes (not always) with OOP it can be helpful to relate our objects with their closest counterparts in the real world.

A Character, or Person, in the real world has attributes, knows things, can do things, etc. No other object should be responsible for storing those, but other objects may need to interact with the Character and as such would need to query or "ask" the Character how they will behave given certain circumstances.

A Game is a human construct and contains rules, logic, calculations, etc. Since this project is about a game, pretty much, it makes sense for the Game object to be all-knowing with respect to the rules and how various objects within the game interact with each other. Careful not to stray too far into making it a "God Object", which is broadly considered an anti-pattern in OOP. That said, "broadly considered" is only one consideration. If your particular application requires a God Object then bloody well make it a God Object. Just be aware of the pros and cons.

A Map in real life is .. well also a human construct but I think what the Map object really relates to in real life is landscape. Landscape knows.. what? Not much except for what it's made of, maybe its own texture and viscosity, weight and other properties like that. A Map Tile in real life is probably just a slice of landscape. So it might be worth considering what exactly these objects have to offer the Game. Likely they can act as a "modifier". Consider this example (and by no means take this as prescriptive) 

1. Game takes user input to move Character
2. Game queries Character for movement ability: Fly? Walk? How far?
3. Game stores coordinates of Character (landscapes don't know this. maybe humans do, but their position in the game is circumstantial to the game?)
4. Game queries Map for the Map Tile at that coordinate
5. Game queries Map Tile for features of the landscape that modify the character's movement through it

I feel like I might have let too much of the cat out of the bag, but it felt like a single example of how one might logic their way through this kind of problem could be helpful. Hopefully it's not overreaching :)

**Comment by ryancrideout** [comment]
Hi Hi YJ,

Looked this over and had some more thoughts:

**Game Engine being a God object -**
 - If we wanted to avoid having a "God object", we could break the engine into multiple parts.
   - I.E., Have a "Math Engine" that handles all of the math, a "Map Engine" that handles all of the map related details, etc.
     - But then would it be alright to have an "Engine of Engines" where we have something to keep track of the engines?

 - I think also that all said, it's hard to imagine the game engine NOT being a God object. Even in my exercise of trying to think of alternatives to having a Game God Engine, it still feels like I wrapped back around to having a "God" object of sorts.

**Maps and Maptiles -**
I still think it makes sense for Maps to keep track of Maptiles as, again, Maps are made up of Maptiles.

That said, I don't think a MapTile should be responsible for knowing who or what is on it - that seems like it's better suited for the game to keep track of those details. I think a MapTile can know what it's made up (E.G., being grass vs dirt), but I don't think it should keep track of it's possible  interactions with whatever resides on the tile.
 - I think my concern about this is that, while I think it makes sense on a "logical" level to keep track of it this way, this does put more responsibility on the Game Engine.

**Characters -**
Reading your thoughts has made me comfortable with the idea of characters knowing their own position and knowing other attributes, but shouldn't be in charge of, say, knowing whether or not they can traverse to a certain location on the map, I think that should be up to the Game Engine.

I think with this mind, the Game Engine would also keep track of the character's position, but then if the Game Engine is responsible for all of the calculations, I don't know if there's much of a point for a character to know their position? I feel like it would be duplicate information at that point.

* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

Am I getting too far lost in the sauce and am missing the point? I think my struggle right now is trying to reconcile the Game Engine being a God Object, because on paper that seems "bad", but as I think on it I really don't see it being any other way.

**Comment by sepiariver** [comment]
> But then would it be alright to have an "Engine of Engines" where we have something to keep track of the engines?

Yes it would be alright, except for those cases when it would not be alright :) Hint: I think in this case it would be fine, good even, but what do you think is an example of a case where it would not be alright?

> hard to imagine the game engine NOT being a God object

Agreed, although to reiterate I'm no game designer

> I think my concern about this is that, while I think it makes sense on a "logical" level to keep track of it this way, this does put more responsibility on the Game Engine.

If the Game Engine is a God Object it has all the responsibilities anyways. If the Game Engine is an Engine of Engines, then some Engine therein will be perfectly suited to handling this particular responsibility. Either way your line of thinking seems to bear fruit here.

>  I don't know if there's much of a point for a character to know their position? I feel like it would be duplicate information at that point.

You're right. Maybe they don't know their own position. A human wandering around a forest with no map and no compass arguably doesn't know anything about their position. So it does not break our model of reality to make the Character unaware of position—only how they can move and other properties innate to the Character.

> my struggle right now is trying to reconcile the Game Engine being a God Object, because on paper that seems "bad", but as I think on it I really don't see it being any other way.

You may (or may not) have missed this line from the previous comment:

> If your particular application requires a God Object then bloody well make it a God Object. Just be aware of the pros and cons.

**Comment by ryancrideout** [comment]
Re: Engine of Engines - Off the top of my head the only case I can think of where it'd be a detriment would be on smaller scale projects where it might not be necessary. 

Like let's say we made a game engine for Checkers - We could make an engine for moving pieces, and another engine for what happens when one piece takes over another, and to top it all off we could have an "Engine of Engines" that handles both of _those_ engines. 

That would work, but alternatively we could just have ONE game engine. Checkers probably isn't so involved to require three separate engines, I'd think one would be good enough. 

Something like that I think would be an argument _against_ an "Engine of Engines", but some may disagree on that.

Either way this was good, thanks for letting me bounce ideas off of you. I think I have some ideas on how to refactor the code so that it more closely aligns with what we were talking about.
